pkl-core/src/main/java/org/pkl/core/runtime/VmUtils.java

```diff
@@ -300,6 +301,59 @@ public static Object doReadMember(
     return result;
   }
 
+  /**
+   * Copies `numberOfLocalsToCopy` locals from `sourceFrame`, starting at `firstSourceSlot`, to
+   * `targetFrame`, starting at `firstTargetSlot`.
+   */
+  public static void copyLocals(
+      VirtualFrame sourceFrame,
+      int firstSourceSlot,
+      VirtualFrame targetFrame,
+      int firstTargetSlot,
+      int numberOfLocalsToCopy) {
+    var sourceDescriptor = sourceFrame.getFrameDescriptor();
+    var targetDescriptor = targetFrame.getFrameDescriptor();
+    // Alternatively, locals could be copied with `numberOfLocalsToCopy`
+    // `ReadFrameSlotNode/WriteFrameSlotNode`'s.
+    for (int i = 0; i < numberOfLocalsToCopy; i++) {
+      var sourceSlot = firstSourceSlot + i;
+      var targetSlot = firstTargetSlot + i;
+      // If, for a particular call site of this method,
+      // slot kinds of `sourceDescriptor` will reach a steady state,
+      // then slot kinds of `targetDescriptor` will too.
+      var slotKind = sourceDescriptor.getSlotKind(sourceSlot);
+      switch (slotKind) {
+        case Boolean -> {
+          targetDescriptor.setSlotKind(targetSlot, FrameSlotKind.Boolean);
+          targetFrame.setBoolean(targetSlot, sourceFrame.getBoolean(sourceSlot));
+        }
+        case Long -> {
+          targetDescriptor.setSlotKind(targetSlot, FrameSlotKind.Long);
+          targetFrame.setLong(targetSlot, sourceFrame.getLong(sourceSlot));
+        }
+        case Double -> {
+          targetDescriptor.setSlotKind(targetSlot, FrameSlotKind.Double);
+          targetFrame.setDouble(targetSlot, sourceFrame.getDouble(sourceSlot));
+        }
+        case Object -> {
+          targetDescriptor.setSlotKind(targetSlot, FrameSlotKind.Object);
+          targetFrame.setObject(
+              targetSlot,
+              sourceFrame instanceof MaterializedFrame
+                  // Even though sourceDescriptor.getSlotKind is now Object,
+                  // it may have been a primitive kind when `sourceFrame`'s local was written.
+                  // Hence we need to read the local with getValue() instead of getObject().
+                  ? sourceFrame.getValue(sourceSlot)
+                  : sourceFrame.getObject(sourceSlot));
+        }
+        default -> {
+          CompilerDirectives.transferToInterpreter();
+          throw new VmExceptionBuilder().bug("Unexpected FrameSlotKind: " + slotKind).build();
+        }
+      }
+    }
```

Comment:
Re: this comment:

> Even though sourceDescriptor.getSlotKind is now Object,
> it may have been a primitive kind when `sourceFrame`'s local was written.
> Hence we need to read the local with getValue() instead of getObject().

I wonder if there is a deeper issue with how we are using frame descriptors. Right now, our `WriteFrameSlotNode` will update the frame descriptor, but I'm not sure if that's totally right. For example, `com.oracle.truffle.api.impl.FrameWithoutBoxing#clear` doesn't update the frame descriptor either.

In any case, you can avoid the weird edge case by looking at the frame slot tag instead:

```suggestion
    for (var i = 0; i < numberOfLocalsToCopy; i++) {
      var sourceSlot = firstSourceSlot + i;
      var targetSlot = firstTargetSlot + i;
      var slotKind = FrameSlotKind.fromTag(sourceFrame.getTag(sourceSlot));
      switch (slotKind) {
        case Boolean -> {
          targetDescriptor.setSlotKind(targetSlot, FrameSlotKind.Boolean);
          targetFrame.setBoolean(targetSlot, sourceFrame.getBoolean(sourceSlot));
        }
        case Long -> {
          targetDescriptor.setSlotKind(targetSlot, FrameSlotKind.Long);
          targetFrame.setLong(targetSlot, sourceFrame.getLong(sourceSlot));
        }
        case Double -> {
          targetDescriptor.setSlotKind(targetSlot, FrameSlotKind.Double);
          targetFrame.setDouble(targetSlot, sourceFrame.getDouble(sourceSlot));
        }
        case Object -> {
          targetDescriptor.setSlotKind(targetSlot, FrameSlotKind.Object);
          targetFrame.setObject(targetSlot, sourceFrame.getObject(sourceSlot));
        }
        default -> {
          CompilerDirectives.transferToInterpreter();
          throw new VmExceptionBuilder().bug("Unexpected FrameSlotKind: " + slotKind).build();
        }
      }
    }
```

Reply:
> One issue I do see here, though, is that we are materializing the same frame multiple times in the case of many generator members in the same for body

Multiple calls to `frame.materialize()` are guaranteed to return the same materialized frame (which is a reference to the mutable virtual frame, not an immutable snapshot thereof). This seems important for Pkl performance because otherwise, every `new X {...}` expression within the same root node would create a new materialized frame.

> Right now, our WriteFrameSlotNode will update the frame descriptor, but I'm not sure if that's totally right.

`WriteFrameSlotNode` updating the frame descriptor is how type profiling of locals works. Note that the slot kind can only change from primitive type to `Object`, i.e., it can only become more general.

> In any case, you can avoid the weird edge case by looking at the frame slot tag instead:

I've considered this but concluded it's not correct. Source frames that share the same descriptor may be copied in an order that differs from the order they were executed/profiled in. If we ignore the source descriptor's slot kind during copying, the target descriptor's slot kind isn't guaranteed to reach a steady state.  For example, long-long-long-Object-Object-Object may turn into Object-long-Object-long-Object-long. As far as I understand, this isn't desirable.

Reply:
> Multiple calls to frame.materialize() are guaranteed to return the same materialized frame (which is a reference to the mutable virtual frame, not an immutable snapshot thereof).

I can't find any docs that guarantee this. Can you provide a reference?

Their own docs say [it returns a new frame](https://www.graalvm.org/truffle/javadoc/com/oracle/truffle/api/frame/Frame.html#materialize()), even if the implementation disagrees.

> I considered this but concluded it's incorrect. Source frames that share the same descriptor may be copied in an order that differs from the order they were executed/profiled in. If we ignore the source descriptor's slot kind during copying, the target descriptor's slot kind isn't guaranteed to reach a steady state. For example, long-long-long-Object-Object-Object may turn into Object-long-Object-long-Object-long. As far as I understand, this isn't desirable.

Ah, I see. That makes sense!

In that case, I don't understand this ternary:

```java
sourceFrame instanceof MaterializedFrame
    ? sourceFrame.getValue(sourceSlot)
    : sourceFrame.getObject(sourceSlot))
```

Why is it safe to call `sourceFrame.getObject` if it is not a `MaterializedFrame`? And, in practice, it's always true; `FrameWithoutBoxing` implements `MaterializedFrame`.

Reply:
> I can't find any docs that guarantee this. Can you provide a reference?

No, but I'm sure it's true. In interpreted code, `frame.materialize()` always returns `frame`. In Graal JITted code, `frame.materialize()` always yields the same `MaterializedFrame` instance. Feel free to double-check by asking in Truffle Slack. 

> Why is it safe to call sourceFrame.getObject if it is not a MaterializedFrame?

Because then `sourceFrame` is a `VirtualFrame`, which means that `copyLocals` is called while `sourceFrame` is active, which guarantees that `sourceFrame.getDescriptor` accurately describes `sourceFrame`.

> And, in practice, it's always true; 

It's always true in interpreted code, where all frames are regular Java objects allocated on the Java heap. Once Graal JIT kicks in, `VirtualFrame` is optimized away (that's what makes it "virtual").